File: UVa/UVa_12455_Bars.py
# http://uva.onlinejudge.org/external/124/12455.pdf

def solution(n, bars):
    p = len(bars)
    for i in range(1 << p):  # 0 ~ 2**n - 1
        sum = 0
        for j in range(p):  # check membership, O(n)
            if i & (1 << j):  # test if bit 'j' is turned on in subset 'i'
                sum += bars[j]
        if sum == n:  # the answer is found
            return "YES"
    return "NO"


def solution2(n, bars):
    from itertools import combinations
    for i in range(len(bars) + 1):
        for c in combinations(bars, i):
            if sum(c) == n:
                return "YES"
    return "NO"


test_case = int(input())
for count in range(test_case):
    n = int(input())
    p = int(input())
    bars = list(map(int, input().split()))
    print(solution2(n, bars))

File: UVa/Problem Solving/cs/UVa_12455_Bars.cpp
// https://onlinejudge.org/external/124/12455.pdf

#include <bits/stdc++.h>
using namespace std;
#define REP(i, n) for (int i = 0; i < n; i++)
#define LSOne(x) ((x) & -(x))

const char *solution(int x, int n, int a[])
{
    for (int i = 0; i < (1 << n); i++)
    {
        int sum = 0;
        int mask = i;
        while (mask)
        {
            int two_pow_j = LSOne(mask);
            int j = __builtin_ctz(two_pow_j);
            sum += a[j];
            mask -= two_pow_j;
        }
        if (sum == x)
            return "YES";
    }
    return "NO";
}

int main()
{
    int t, n, p;
    int a[1000];

    scanf("%d", &t);
    while (t--)
    {
        scanf("%d %d", &n, &p);
        REP(i, p)
        {
            scanf("%d", &a[i]);
        }
        printf("%s\n", solution(n, p, a));
    }

    return 0;
}